---
title: "PERFORMANCE REPORT AGE 42"
output:
  flexdashboard::flex_dashboard:
    output_dir: docs
    orientation: rows
    vertical_layout: scroll
    source_code: embed
    css: estilo.css
    mathjax: NULL
    self_contained: FALSE
    logo: "Siwa-blanco-01.png"
    #runtime: shiny
    #navbar:
     # - { title: "About SIWA", href: "https://siwa.bio/" }
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r libraries, include=FALSE, cache=TRUE}
###load libraries panels tabs####
library(phyloseq) #
library(DESeq2) #
library(kableExtra) #
library(genefilter) #
library(microbiome) #
library(ggplot2)  #
library(vegan) #
library(ggpubr) #
library(ggplot2) #
library(plyr)
library(multcompView)

#####Load libraries correlations####
library(ggplot2)
library(phyloseq)
library(stringr)
library(dplyr)

#####Load libraries LR####
library(tidyverse)
library(ape)
library(ggpubr)

####Load libraries categorical histo####
library(kableExtra)
library(plyr)
#library("fantaxtic")
library(data.table)

####load libraries Ratios-histo boxplots####
library(multcompView)
library(reshape)

### OTRICAS
library(plotly)
library(plyr)
library(flexdashboard)
library(shiny)
library(DT)
library(stringr)
library(rstatix)
#library(microViz)
library(lsr)
library(readxl)
library(openxlsx)

## FUNCTIONS
source("/Users/sebastianbedoyamazo/Documents/siwa_git/Methods-review/functions.R")
source("/Users/sebastianbedoyamazo/Documents/siwa_git/Methods-review/correlations.R")
```




```{r All inputs SEBAS, include=FALSE}
#open phyloseq object
input_dir = "/Users/sebastianbedoyamazo/Documents/siwa_git/Performance/inputdata/"

ODLEPobj <- readRDS(paste0(input_dir, "PhyloseqObject_perf.rds"))

## Data creada con "" - Jupyter Notebook
metadata <- read_excel("//Users/sebastianbedoyamazo/Documents/siwa_git/Performance/inputdata/metadata_test.xlsx")


```

```{r open data 1, include=FALSE}
#Organize metadata
# metadata <- metadata %>% rename(SampleID = sampleid, AnimalID = animalid...4, SampleLocation = samplelocation, Treatment = Trt)
# metadata <- metadata %>% 
#   tibble::column_to_rownames(var = "SampleID") %>% 
#   tibble::add_column(SampleID = rownames(.), .before = 1)
# metadata <- metadata[order(metadata$SampleID), ]

```

```{r open data 1, include=FALSE}
#Extract taxonomy and otu table from phyloseq object
taxonomy <-as.data.frame(tax_table(ODLEPobj))
taxonomy$OTU <- row.names(taxonomy)  
otu_table <- as.data.frame(otu_table(ODLEPobj))
otu_table$OTU<- row.names(otu_table)
metadata <- as.data.frame(sample_data(ODLEPobj))
meta_exp <- metadata
# is.data.frame(metadata)

# ### AGGREGATE BY TAXA
pseq.rel <- microbiome::transform(ODLEPobj, "compositional")
phyloseq_relative <- pseq.rel
colnames(taxonomy)
grouped <- aggregate_taxa_siwa(phyloseq_relative, "family")
  



samplesE267 <- unlist(meta_exp[meta_exp$projectid == 'E267', 'SampleID'])
samplesE271 <- unlist(meta_exp[meta_exp$projectid == 'E271', 'SampleID'])
samplesE345 <- unlist(meta_exp[meta_exp$projectid == 'E345', 'SampleID'])
samplesE347 <- unlist(meta_exp[meta_exp$projectid == 'E347', 'SampleID'])


meta_exp <- as.matrix(meta_exp)
meta_exp <- as.data.frame(meta_exp)
class(meta_exp)

meta_expE267 <- filter(meta_exp, projectid == "E267")
meta_expE271 <- filter(meta_exp, projectid == "E271")
meta_expE345 <- filter(meta_exp, projectid == "E345")
meta_expE347 <- filter(meta_exp, projectid == "E347")




#define dataframe for analysis as otu_table
otu_table<-grouped[,samplesE347]
#otu_table <- grouped
meta_exp <- meta_expE347
#Transpose the data to have sample names on rows
otu_table <- as.matrix(otu_table)
otu_table <- t(otu_table)
class(otu_table)

rownames(otu_table) <-
  sapply(str_remove_all(rownames(otu_table), "X"), "[") # porsi

#rownames(otu_table) #deben ser las samples
```


```{r transformation 1, include=FALSE}
#####transformation and filter for compositional data
# calcula la proporción de ceros en cada columna de un data frame df. La función apply() se utiliza para aplicar una función a cada columna del data frame. La función que se aplica es df == 0, que compara cada valor en el data frame con cero y devuelve un vector lógico que indica si cada valor es igual a cero o no. El resultado de esta comparación es una matriz lógica con el mismo número de filas y columnas que el data frame.
# 
# Luego, se utiliza la función sum() junto con el argumento 2 de apply() para sumar los valores en cada columna de la matriz lógica, lo que da como resultado el número total de ceros en cada columna. Dividiendo este resultado por el número de filas en el data frame dim(df)[1], se obtiene la proporción de ceros en cada columna.
# 
# Finalmente, se asigna el resultado a un objeto llamado rowz. Por lo tanto, rowz será un vector que contiene la proporción de ceros en cada columna del data frame df.

rowz<-apply(otu_table == 0, 2, sum)/dim(otu_table)[1] #sum over columns (OTUS)
# dejar otus frecuentes, quitando los que están llenos de CEROS
p <- which(rowz>0.99)
otu_table_filtered <-otu_table[, -p]


# colz <- apply(otu_table_filtered == 0, 1, sum)/dim(otu_table_filtered)[2]
#pp<-which(colz>0.9)
# colnames(otu_table)
# otu_table[,"UNKNOWN"]
#  as.matrix() es una función utilizada para convertir un objeto en una matriz. En este caso, la línea de código df <- as.matrix(df) convierte el objeto df en una matriz y lo asigna de vuelta a la variable df.
# 
# La función as.matrix() es útil cuando se desea trabajar con una matriz en lugar de un data frame u otro tipo de objeto. Al convertir un data frame en una matriz, se pierden algunas de las características específicas de los data frames, como los nombres de columna y fila, pero se obtiene una estructura de datos más simple que se puede manipular más fácilmente con algunas funciones de R.
# 
# Es importante tener en cuenta que la función as.matrix() crea una copia de los datos originales, por lo que si el data frame es grande, puede requerir una cantidad significativa de memoria adicional para almacenar la matriz. Además, si el data frame original tiene columnas de diferentes tipos de datos, la función as.matrix() puede forzar todos los datos a un solo tipo, lo que puede no ser deseable en algunas situaciones.

otu_table_filtered<- as.matrix(otu_table_filtered)
require("zCompositions")
require("compositions")


# cmultRepl() para transformar el data frame df en una matriz de composición. Esta función multiplica cada valor en el data frame por un factor constante de modo que la suma de todos los valores en cada fila sea igual a 1. La matriz resultante tiene una fila para cada observación y una columna para cada variable en el data frame

otu_table_filtered_transformed <- cmultRepl(otu_table_filtered)

# la función clr() para transformar la matriz de composición resultante en una matriz de log-ratios utilizando el método log-ratio centrado. Esta transformación es comúnmente utilizada en análisis de composición para transformar los datos en un espacio lineal que se puede utilizar en análisis estadísticos estándar.

#otu_table_filtered_transformed_log = otu_table_filtered_transformed
otu_table_filtered_transformed_log <- clr(otu_table_filtered_transformed)

dim(otu_table_filtered_transformed_log)



#Extract the corresponding meta_table for the samples in abund_table
#rownames(meta_exp) <- meta_exp$sampleid
#meta_exp <- meta_exp %>% column_to_rownames(var = "sampleid")
meta_exp <- meta_exp[rownames(otu_table_filtered_transformed_log), ]
rownames(otu_table_filtered_transformed_log)
colnames(meta_exp)

```





```{r choose variables 1, include=FALSE}

#When its ony one variable y just leave the c("") with onw variable.
#You can use sel_env to specify the variables you want to use and sel_env_label to specify the labels for the pannel

sel_vars <-
  c(
    "BW42",
    #"FCR0_42",
    "FCR36_42",
    "BWG0_42",
    "BWG36_42"
  )

## Asi van a aparecer en el plot
sel_vars_label <- list(
  "BW42" = "BW (42 days)",
    #"FCR0_42" = "FCR (0-42 days)",
    "FCR36_42" = "FCR (36-42 days)",
    "BWG0_42" = "BWG (0-42 days)",
    "BWG36_42" = "BWG (36-42 days)"
  
)

sel_vars_label<-t(as.data.frame(sel_vars_label))
sel_vars_label<-as.data.frame(sel_vars_label)
colnames(sel_vars_label)<-c("Trans")


#Now get a filtered table based on sel_env --> DEJAR SOLO SAMPLES CON METADATOS
meta_exp_filtered<-meta_exp[,sel_vars]
meta_exp_filtered <- as.data.frame(meta_exp_filtered)

X <- otu_table_filtered_transformed_log
X <- X[, order(colSums(X), decreasing = TRUE)]  #### REVISAAAAR CENTERED LOG RATIO TRANSFORM!!!! 
dim(X) 

#Extract list of top N Taxa
# N<-20
# otus_list<-colnames(X)[1:56]
otus_list<-colnames(X)


X <- data.frame(X[, colnames(X) %in% otus_list], check.names = FALSE)
Y <- meta_exp_filtered
dim(X)
dim(Y)
#Get grouping information
groups_ <- meta_exp$SampleLocation
# #### Convertir las variables a numéricas
str(Y)
Y$BW42 <- as.numeric(Y$BW42 )
#Y$FCR0_42 <- as.numeric(Y$FCR0_42)
Y$FCR36_42 <- as.numeric(Y$FCR36_42)
Y$BWG0_42 <- as.numeric(Y$BWG0_42)
Y$BWG36_42 <- as.numeric(Y$BWG36_42)

```



```{r correlation method 1, include=FALSE}
corr_347_family =  correlation_siwa(X, Y, groups_,  method="pearson")


```

```{r correlation method 1, include=FALSE}


#correlation by Species aggregate 
corr_267_specie$projectid <- "E267"
corr_271_specie$projectid <- "E271"
corr_345_specie$projectid <- "E345"
corr_347_specie$projectid <- "E347"

corr_analysis_species = rbind(corr_267, corr_271, corr_345, corr_347)
corr_analysis_species$Taxa_aggre = "species"

#correlation by Genus aggregate 
corr_267_genus$projectid <- "E267"
corr_271_genus$projectid <- "E271"
corr_345_genus$projectid <- "E345"
corr_347_genus$projectid <- "E347"

corr_analysis_genus = rbind(corr_267_genus, corr_271_genus, corr_345_genus, corr_347_genus)
corr_analysis_genus$Taxa_aggre = "genus"

#correlation by Family aggregate 
corr_267_family$projectid <- "E267"
corr_271_family$projectid <- "E271"
corr_345_family$projectid <- "E345"
corr_347_family$projectid <- "E347"

corr_analysis_family = rbind(corr_267_family, corr_271_family, corr_345_family, corr_347_family)
corr_analysis_family$Taxa_aggre = "family"

colnames(corr_analysis_species)
colnames(corr_analysis_genus)
colnames(corr_analysis_family)


corr_analysis = rbind(corr_analysis_species, corr_analysis_genus, corr_analysis_family)


corr_analysis_filter <-  corr_analysis[corr_analysis$AdjPvalue < 0.05, ]

#write.xlsx(corr_analysis, "/Users/sebastianbedoyamazo/Documents/siwa_git/Performance/inputdata/corr_analysis.xlsx")


Taxa_corr_unique <- as.data.frame(unique(corr_analysis_filter$Taxa))

```


```{r correlation method 1, include=FALSE}
corr_analysis_filter_taxa_aggre <- split(corr_analysis_filter, f = corr_analysis_filter$Taxa_aggre)
corr_analysis_filter_taxa_aggre

```


```{r correlation method 1, include=FALSE}
corr_analysis_filter_taxa_aggre_family <- split(corr_analysis_filter_taxa_aggre$family, f = corr_analysis_filter_taxa_aggre$family$Type)
corr_analysis_filter_taxa_aggre_family

```



```{r correlation method 1, include=FALSE}
corr_analysis_filter_taxa_aggre_genus <- split(corr_analysis_filter_taxa_aggre$genus, f = corr_analysis_filter_taxa_aggre$genus$Type)
corr_analysis_filter_taxa_aggre_genus

```
```{r correlation method 1, include=FALSE}
corr_analysis_filter_taxa_aggre_species <- split(corr_analysis_filter_taxa_aggre$species, f = corr_analysis_filter_taxa_aggre$species$Type)
corr_analysis_filter_taxa_aggre_species

```





```{r correlation method 1, include=FALSE}

# corr_ileum <- split(corr_analysis_filter$ileum, f = ifelse(corr_analysis_filter$ileum$Correlation >= 0, "positive_cor", "negative_cor"))

```



```{r crear plot 1, include=TRUE,echo=FALSE, warning=FALSE, fig.dim = c(10, 6)}
#scatterplot most corr and significant taxa

p1 <- ggplot(merged_df, aes(x = Lactobacillus_aviarius, y = BWG0_42, col = projectid))+
  geom_smooth(method='lm', se=FALSE)+
  facet_wrap(~SampleLocation) + 
  labs(x = "Relative abundace Lactobacillus_aviarius (no transformation data)",y = "BW (42 days)")+
  geom_point()+
  theme(legend.position="right")

p2 <- ggplot(merged_df_clr, aes(x = Lactobacillus_aviarius, y = BWG0_42, col = projectid))+
  geom_smooth(method='lm', se=FALSE)+
  facet_wrap(~SampleLocation) + 
  labs(x = "Relative abundace Lactobacillus_aviarius (CLT)",y = "BW (42 days)")+
  geom_point()+
  theme(legend.position="right")




p1
p2


```

```{r crear plot 1, include=TRUE,echo=FALSE, warning=FALSE, fig.dim = c(10, 6)}
#data otu_table

p1 <- ggplot(merged_df, aes(x = Lactobacillus_aviarius, y = BWG36_42, col = projectid))+
  geom_smooth(method='lm', se=FALSE)+
  facet_wrap(~SampleLocation) + 
  labs(x = "Relative abundace Lactobacillus_aviarius (no transformation data)",y = "BWG (36-42 days)")+
  geom_point()+
  theme(legend.position="right")

p2 <- ggplot(merged_df_clr, aes(x = Lactobacillus_aviarius, y = BWG36_42, col = projectid))+
  geom_smooth(method='lm', se=FALSE)+
  facet_wrap(~SampleLocation) + 
  labs(x = "Relative abundace Lactobacillus_aviarius (CLT)", y = "BWG (36-42 days)")+
  geom_point()+
  theme(legend.position="right")




p1
p2


```


```{r crear plot 1, include=TRUE,echo=FALSE, warning=FALSE, fig.dim = c(10, 6)}
#data otu_table

p1 <- ggplot(merged_df, aes(x = Lactobacillus_aviarius, y = FCR0_42, col = projectid))+
  geom_smooth(method='lm', se=FALSE)+
  facet_wrap(~SampleLocation) + 
  labs(x = "Relative abundace Lactobacillus_aviarius (no transformation data)",y = "FCR (0-42 days)")+
  geom_point()+
  theme(legend.position="right")

p2 <- ggplot(merged_df_clr, aes(x = Lactobacillus_aviarius, y = FCR0_42, col = projectid))+
  geom_smooth(method='lm', se=FALSE)+
  facet_wrap(~SampleLocation) + 
  labs(x = "Relative abundace Lactobacillus_aviarius (CLT)",y = "FCR (0-42 days)")+
  geom_point()+
  theme(legend.position="right")




p1
p2


```



```{r crear plot 1, include=TRUE,echo=FALSE, warning=FALSE, fig.dim = c(10, 6)}
#data otu_table

p1 <- ggplot(merged_df, aes(x = Lactobacillus_aviarius, y = FCR36_42, col = projectid))+
  geom_smooth(method='lm', se=FALSE)+
  facet_wrap(~SampleLocation) + 
  labs(x = "Relative abundace Lactobacillus_aviarius (no transformation data)",y = "FCR (36-42 days)")+
  geom_point()+
  theme(legend.position="right")

p2 <- ggplot(merged_df_clr, aes(x = Lactobacillus_aviarius, y = FCR36_42, col = projectid))+
  geom_smooth(method='lm', se=FALSE)+
  facet_wrap(~SampleLocation) + 
  labs(x = "Relative abundace Lactobacillus_aviarius (CLT)",y = "FCR (36-42 days)")+
  geom_point()+
  theme(legend.position="right")




p1
p2


```





### FIGURE 16: Kendall correlation heatmap between microbial families and host genes. <br/> Families such as Sutterellaceae, Erysipelotcrichaceae, Eggerthellaceae, and Anaerovoracaceae are strongly associated with higher Il10 expression in the ileum.  


```{r correlation method 1, include=FALSE}
df <- filter(corr_analysis_filter, Taxa_aggre == "species")


Env_labeller <- function(variable,value){
  return(sel_vars_label[as.character(value),"Trans"])
}

df$label <- df$Taxa

method="pearson"

```


```{r crear plot 1, include=TRUE,echo=FALSE, warning=FALSE, fig.dim = c(10, 6)}

p <- ggplot(aes(x = Type, y = Taxa, fill = Correlation), data = df)

p <-
  p + geom_tile() + scale_fill_gradient2(low = "#2C7BB6", mid = "white", high =
                                           "#D7191C")
p <-
  p + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))

p <-
  p + geom_text(aes(label = Significance), color = "black", size = 3) + labs(y =
                                                                               NULL, x = NULL, fill = method)
p <-
  p + facet_grid(
    . ~ Env,
    drop = TRUE,
    scale = "free",
    space = "free_x",
    labeller = Env_labeller
  )+ theme(text=element_text(size=9))
  
p
```

